pandas/util/_exceptions.py

```diff
@@ -25,20 +26,24 @@ def rewrite_exception(old_name: str, new_name: str) -> Iterator[None]:
         raise
 
 
-def find_stack_level() -> int:
+@functools.lru_cache
+def find_stack_level(frame) -> int:
     """
     Find the first place in the stack that is not inside pandas
     (tests notwithstanding).
+
+    ``frame`` should be passed as ``inspect.currentframe()`` by the
+    calling function.
+
+    https://stackoverflow.com/questions/17407119/python-inspect-stack-is-slow
     """
 
     import pandas as pd
 
     pkg_dir = os.path.dirname(pd.__file__)
     test_dir = os.path.join(pkg_dir, "tests")
 
-    # https://stackoverflow.com/questions/17407119/python-inspect-stack-is-slow
-    frame = inspect.currentframe()
-    n = 0
```

Comment:
Hmmm. This is causing a lot of code churn. 
What if we did something like
```
@functools.lru_cache
def _find_stack_level_inner(currentframe):
     ....
def find_stack_level():
      frame = inspect.currentframe()
      ...
      # We are adding an extra stack level in wrapping the "real" find stack level function
      # need to remove
      return _find_stack_level_inner(frame) - 1
```

where _find_stack_level_inner is everything below this line.

Reply:
hadn't thought of that - will try, thanks!

Reply:
Just tried that, because then the value of `frame` is different each time the function is called and so caching doesn't work

Reply:
E.g. if I run
```python
import inspect

def print_currentframe(frame = None):
    if frame is None:
        frame = inspect.currentframe()
    print(frame)

for _ in range(3):
    print_currentframe()

for _ in range(3):
    print_currentframe(inspect.currentframe())
```
then it outputs
```
<frame at 0x70c40acd60, file '/home/megorelli/pandas-dev/t4.py', line 6, code print_currentframe>
<frame at 0x70c4184420, file '/home/megorelli/pandas-dev/t4.py', line 6, code print_currentframe>
<frame at 0x70c41163c0, file '/home/megorelli/pandas-dev/t4.py', line 6, code print_currentframe>
<frame at 0xdbe2f40, file '/home/megorelli/pandas-dev/t4.py', line 12, code <module>>
<frame at 0xdbe2f40, file '/home/megorelli/pandas-dev/t4.py', line 12, code <module>>
<frame at 0xdbe2f40, file '/home/megorelli/pandas-dev/t4.py', line 12, code <module>>
```

Note how the first 3 differ (so no caching would take place), but the last 3 are the same (so caching would work)

Reply:
Ah, ok I see. nevermind then.